eth/blockwatch/stack_test.go

```diff
@@ -0,0 +1,317 @@
+package blockwatcher
+
+import (
+	"errors"
+	"sync"
+	"testing"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type stubMiniHeaderStore struct {
+	headers           []*MiniHeader
+	latestErr         error
+	sortedByNumberErr error
+	insertErr         error
+	deleteErr         error
+}
+
+func (s *stubMiniHeaderStore) FindLatestMiniHeader() (*MiniHeader, error) {
+	if s.latestErr != nil {
+		return nil, s.latestErr
+	}
+
+	if len(s.headers) == 0 {
+		return nil, nil
+	}
+
+	return s.headers[len(s.headers)-1], nil
+}
+
+func (s *stubMiniHeaderStore) FindAllMiniHeadersSortedByNumber() ([]*MiniHeader, error) {
+	if s.sortedByNumberErr != nil {
+		return nil, s.sortedByNumberErr
+	}
+
+	return s.headers, nil
+}
+
+func (s *stubMiniHeaderStore) InsertMiniHeader(header *MiniHeader) error {
+	if s.insertErr != nil {
+		return s.insertErr
+	}
+
+	s.headers = append(s.headers, header)
+
+	return nil
+}
+
+func (s *stubMiniHeaderStore) DeleteMiniHeader(hash ethcommon.Hash) error {
+	if s.deleteErr != nil {
+		return s.deleteErr
+	}
+
+	for i, header := range s.headers {
+		if header.Hash == hash {
+			copy(s.headers[i:], s.headers[i+1:])
+			s.headers[len(s.headers)-1] = nil
+			s.headers = s.headers[:len(s.headers)-1]
+			return nil
+		}
+	}
+
+	return errors.New("MiniHeader not found")
+}
+
+func TestPop(t *testing.T) {
+	store := &stubMiniHeaderStore{}
+	stack := NewStack(store, 10)
+
+	assert := assert.New(t)
+	require := require.New(t)
+
+	// Test when store.FindLatestMiniHeader() returns error
+	store.latestErr = errors.New("FindLatestMiniHeader error")
+	_, err := stack.Pop()
+	assert.EqualError(err, store.latestErr.Error())
+
+	// Test when store returns nil MiniHeader
+	store.latestErr = nil
+	header, err := stack.Pop()
+	assert.Nil(header)
+	assert.Nil(err)
+
+	// Test when store.DeleteMiniHeader() returns error
```

Comment:
Just a detail but since `DeleteMiniHeader` is a stub where we error simply if `store.deleteErr` is set on the `store` struct I would place this comment at line 90 where we are actually setting `store.deleteErr`. The current placement of the comment where we are simply adding a first header to the store had me confused for a second as that doesn't lead to an error being returned.

Reply:
Fixed in 9b746c0